### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by marco-michel**

### CODE REVIEW

File: core/krylovObservables.cpp [hunk 1]
```diff
@@ -1,50 +1,238 @@
...
+}
+
+
+/**
+ * Writes sampled expectation values of a list of observables to file
+ * @param obs_list List of observables with sampled expectation values
+ * @param para List of parameters that are included as attributes in the HDF5 file as well as filename (if requested)
+ * @param name Requested filename
+ */
+void krylovBasicObservable::saveResult(const std::vector<std::unique_ptr<krylovBasicObservable>> &obs_list,  parameter_list &para, const std::string &name)
```
[the remark this answers is not in the document]
Comment (sebastian-zell) [reply core/krylovObservables.cpp:61]: Or even better may be to resurrect some of the krylovHelper, i.e. a class in the helper-folder that is given an observable and then outputs it. This has the additional advantage that the core is independent of HDF5. 
But let's see what you think. :)

File: helper/hamiltonian.cpp [hunk 2]
```diff
@@ -281,30 +276,28 @@ std::string Hamiltonian::toString()
...
-void Hamiltonian::createHamiltonMatrix(smatrix * &out, basicBasis* basis)
+std::unique_ptr<smatrix> Hamiltonian::createHamiltonMatrix(basicBasis* basis)
 {
-	out = createMatrix(hamiltonOperator, basis);
+	return createMatrix(hamiltonOperator, basis);
 }
 
 /**
  * Creates an array of observables (represented as matrices). For each mode, an observable corresponding to the expectation value of the of the occupation of the mode is created.
  * @param out resulting list of observables
```
[the remark this answers is not in the document]
Comment (marco-michel) [reply helper/hamiltonian.cpp:286]: 9477e0a

File: helper/hamiltonian.cpp [hunk 3]
```diff
@@ -281,30 +276,28 @@ std::string Hamiltonian::toString()
  * @param out Resulting sparse matrix
```
[the remark this answers is not in the document]
Comment (marco-michel) [reply helper/hamiltonian.cpp]: 9477e0a

File: helper/hamiltonian.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (marco-michel) [reply helper/hamiltonian.cpp]: 5a9394e

File: helper/Basis.h [hunk 4]
```diff
@@ -178,6 +180,10 @@ struct basisVector
 		return true;
 	}
 
+	/**
+	* Assign operator
+	* @param b BasisVector from which a copy is made
```
[the remark this answers is not in the document]
Comment (marco-michel) [reply helper/Basis.h]: 1a1896bd3075b309686aa6d573ef627a155fa5d0

File: core/matrixDataTypes.cpp [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+    for (size_t i = 0; i < numValues; i++) {
+        out[rowIndex[i]] += alpha * values[i] * in[columns[i]];
+    }
+    return 0;
+#endif
+}
+
+/**
+* Setting up library variables to speed up spMV. Or do nothing if none is installed.
+*/
```
[the remark this answers is not in the document]
Comment (marco-michel) [reply core/matrixDataTypes.cpp:237]: e5b0688e1cb7b59da8abdd7becd4cee27ac6eb12

File: core/krylovLogger.h [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+	enum loggingLevel { DEBUG, INFO, WARNING, ERROR, FATAL };
+
+	krylovLogger() = default;
+	krylovLogger(loggingLevel level);
+	void set_loggingLevel(loggingLevel level);
+	void log_message(loggingLevel level, const std::string& msg);
+	static std::string stringStreamWrapperd(double val);
+
+private:
+	bool logToFile;
```
[the remark this answers is not in the document]
Comment (marco-michel) [reply core/krylovLogger.h]: It is not used at the moment. I started but then postponed the possibility to log to a file instead of terminal output. It is not much work, but I didn't want to delay the new TimeEvolver release for that. I removed the variable (and its initialization) for now. 
c4f05ab6

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,5 +1,33 @@
 # Changelog
 
+## 2.0.0 - --
+
+### Added
+- Implemented a rudimentary logger (based on BOOST log) including a verbose option. Logging to file will be supported in a later release. 
+- To ensure data ownership and avoiding the copying of large matrices all large matrices including observables are now expected to be passed as unique_ptr.
```
[the remark this answers is not in the document]
Comment (marco-michel) [reply CHANGELOG.md]: d27a741524fdee7499f043474b96ca95b2551a87

File: core/krylovObservables.cpp [hunk 8]
```diff
@@ -1,50 +1,238 @@
...
+	return obs_name;
+}
 
-#include "matrixDataTypes.h"
-#include "krylovObservables.h"
+/**
+* Function to return a pointer to the expectation value array
+* @return Raw pointer to the array which stores the expectation values
+*/
+double* krylovBasicObservable::retexpectationValues()
```
[the remark this answers is not in the document]
Comment (marco-michel) [reply core/krylovObservables.cpp]: 2b70bd10

File: core/krylovObservables.cpp [hunk 9]
```diff
@@ -1,50 +1,238 @@
...
+{
+    return expectationValues;
+}
 
+/**
+* Function to return the number of samples taken
+* @return Number of samples taken
+*/
+size_t krylovBasicObservable::retnumSamples()
```
[the remark this answers is not in the document]
Comment (marco-michel) [reply core/krylovObservables.cpp]: 2b70bd10

File: core/krylovTimeEvolver.h [hunk 10]
```diff
@@ -1,96 +1,101 @@
...
-
+/**
+* Main class responsible for computing the actual time evolution.
+*/
 class krylovTimeEvolver
 {
 public:
-    krylovTimeEvolver(double t, size_t Hsize, std::complex<double>* v, double samplingStep, double tol, int mm, smatrix** observables, int nbObservables, smatrix* Ham, std::complex<double> expFactor, bool checkNorm= true, bool fastIntegration = false);
-    krylovTimeEvolver(double t, size_t Hsize, std::complex<double>* v, double samplingStep, double tol, int mm, std::vector<std::unique_ptr<krylovBasicObservable>>  observables, smatrix* Ham, std::complex<double> expFactor, bool checkNorm = true, bool fastIntegration = false, bool progressBar = false);
+    krylovTimeEvolver(double t, std::complex<double>* v, double samplingStep, double tol, int mm, std::vector<std::unique_ptr<krylovBasicObservable>>  observables, std::unique_ptr<smatrix> HamIn, double expFactor = 1.0, bool fastIntegration = false, bool progressBar = true);
```
[the remark this answers is not in the document]
Comment (marco-michel) [reply core/krylovTimeEvolver.h:59]: You are right. I kept the order of arguments for legacy reasons but since we changed the datatypes anyway this does not make much sense. 
However, I am not a huge fan of setting so many standard values there. I don't know if it is even possible since calling it (with all standard values set) would be identical to the short one. There would be no way for the compiler to tell which constructor is meant. On top of that, getting the order of importance right would be impossible. The only benefit it would give is that "standard" values would be apparent from the constructor signature. 
I would prefer to mention standard values in the constructor of the short version (or/also somewhere else). 

In summary, I see 2 options. 
1) Kill the second short constructor and set default values for all the values the short constructor does it. 
2a) Keep the long one with an intermediate number of default values
2b) Remove default values in the constructor altogether and mention standard values. (This is the version I would prefer. I spend quite some time in finding apparent errors caused by me messing up the order of default parameters....)

File: core/krylovTimeEvolver.cpp [hunk 11]
```diff
@@ -142,75 +108,33 @@ krylovTimeEvolver::~krylovTimeEvolver()
...
-constexpr std::complex<double> krylovTimeEvolver::zero;
+
 
 /**
  * Computes and saves values of observables for current state ('sampledState')
  */
-void krylovTimeEvolver::sample()
-{
-	if (progressBar)
+void krylovTimeEvolver::sample() {
```
[the remark this answers is not in the document]
Comment (marco-michel) [reply core/krylovTimeEvolver.cpp:120]: Besides the unused variable i :)
d6dc820f

File: core/krylovTimeEvolver.cpp [hunk 12]
```diff
@@ -704,6 +645,45 @@ void krylovTimeEvolver::printProgress(float prog)
...
+		}
+	std::this_thread::sleep_for(std::chrono::milliseconds(500));
+	}
+	return;
+}
+
+
+/**
+* Initialize member variables of return structure
+*/
```
[the remark this answers is not in the document]
Comment (marco-michel) [reply core/krylovTimeEvolver.cpp:669]: 5879b3f5

File: core/krylovTimeEvolver.cpp [hunk 13]
```diff
@@ -387,21 +303,37 @@ krylovReturn* krylovTimeEvolver::timeEvolve()
...
 	//The (m+1,m) element of Hessenberg matrix (needed for computation of error)
 	double h = 0;
 	//Eigenvalues of Hessenberg matrix
-	std::complex<double> *eigenvalues = new std::complex<double>[m];
+	std::complex<double>* eigenvalues = new std::complex<double>[m];
 	//Eigenvectors of Hessenberg matrix
-	std::complex<double> *schurvector = new std::complex<double>[m * m];
+	std::complex<double>* schurvector = new std::complex<double>[m * m];
+
+	//Init observable vector before first sample
```
[the remark this answers is not in the document]
Comment (marco-michel) [reply core/krylovTimeEvolver.cpp]: 4a920d9e

File: core/krylovTimeEvolver.cpp [hunk 14]
```diff
@@ -387,21 +303,37 @@ krylovReturn* krylovTimeEvolver::timeEvolve()
...
-	sample();
 
+
+	//Start progressBar thread
+	if (progressBar == true)
+			pBThread = std::thread(&krylovTimeEvolver::progressBarThread, this);
+	
     //Main loop
     while (index_samples < n_samples)
     {
```
[the remark this answers is not in the document]
Comment (marco-michel) [reply core/krylovTimeEvolver.cpp:333]: Who knows :D? 
It is there since 1.0 apparently, but with 2.0 it will be finally removed. 
806d664c

File: core/krylovTimeEvolver.cpp [hunk 15]
```diff
@@ -595,17 +527,20 @@ krylovReturn* krylovTimeEvolver::timeEvolve()
...
 	cblas_zcopy(Hsize, currentVec, 1, VRet->values, 1);
 
 	for (size_t j = 0; j <= m - 1; j++) {
-		sparse_status_t mklStatus = mkl_sparse_z_mv(SPARSE_OPERATION_NON_TRANSPOSE, expFactor, *HamOpt,
-				descriptor, (VRet->values) + j * Hsize, zero, tmpBlasVec);
-        if(SPARSE_STATUS_SUCCESS != mklStatus)
+
+		int spStatus = Ham->spMV(expFactor, (VRet->values) + j * Hsize, tmpBlasVec);
+
+		//sparse_status_t mklStatus2 = mkl_sparse_z_mv(SPARSE_OPERATION_NON_TRANSPOSE, expFactor, *HamOpt, descriptor, (VRet->values) + j * Hsize, zero, tmpBlasVec);
```
[the remark this answers is not in the document]
Comment (marco-michel) [reply core/krylovTimeEvolver.cpp]: Yes! Just kept there to check what is supposed to happen there. I forgot to remove it ...
a1a3cd7a

File: core/krylovTimeEvolver.cpp [hunk 16]
```diff
@@ -494,46 +427,58 @@ krylovReturn* krylovTimeEvolver::timeEvolve()
...
+			} catch (requestStopException& e)
+			{
+				logger.log_message(krylovLogger::ERROR, e.what());
+				delete[] eigenvalues; delete[] schurvector;
+				delete V; delete H;
+				if (progressBar){
+					stop_printing = true;
+					pBThread.join();
+				}
+				statusCode = 1000;
```
[the remark this answers is not in the document]
Comment (marco-michel) [reply core/krylovTimeEvolver.cpp]: Yes, definitely. I didn't think too much about it. As you pointed out, the time evolution is perfectly alright up to this point (and would be most likely also after that).  If observables have a different measure of success it's not of TimeEvolver business.
Besides of the loss of previous errors/warnings in that case. This will be addressed in some future release when we rework the statusCode and error handling. 

I assigned the creative number 3 then to this case for now. 
41434eb5

File: core/version.h [hunk 17]
```diff
@@ -0,0 +1,9 @@
+#pragma once
+
+
+//	TIMEEVOLVER_VERSION / 100 is major version
```
[the remark this answers is not in the document]
Comment (marco-michel) [reply core/version.h:4]: I saw the boost library doing it that way. 
It describes how you get the integer value of the major and minor version number. 
E.g.: Version 2.11 would be represented as TIMEEVOLVER_VERSION=211.
You can get the major version 2 as 2 = TIMEEVOLVER_VERSION  / 100
and the minor version number 11 as TIMEEVOLVER_VERSION % 100 (modulo/ division with remainder)

File: core/krylovObservables.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (marco-michel) [reply core/krylovObservables.cpp:61]: The class parameter is defined in parameter.h which is located in the helper folder. 
I think we agree that krylovObservable should be in core since the the timeEvolver expects one of those as argument. 
Parameter is in helper since although it appears in the argument of saveResult (core) it is completely optional. 
On a side note, krylovObservable is strictly speaking also optional, the timeEvolver will just not output a lot when it's missing (besides the final state).   

We could outsource the saveResult function, which is currently a static class function of krylovObservable, to a separate file in the helper folder. It should not be much work (since it is already a static function). However, I don't think it is an improvement. I also don't have any problems if there is HDF5 code in the core :)

But I am still open to be convinced about it :)

File: core/krylovTimeEvolver.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (sebastian-zell) [reply core/krylovTimeEvolver.h:59]: Sounds good, let's do 2b then. Still with reordering of arguments, right?

File: core/version.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (sebastian-zell) [reply core/version.h:4]: sounds fine :)

File: core/krylovTimeEvolver.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (marco-michel) [reply core/krylovTimeEvolver.h]: cd75537
